src/glb-director-xdp/main.go

```diff
@@ -340,26 +341,86 @@ func (app *Application) ReloadForwardingTable() {
 	}
 }
 
+func gracefullReloadByExec() {
+	fmt.Printf("Reloading by exec-ing a new version of glb-director-xdp\n")
+
+	// give us a temp working dir for socket comm
+	tmpDir, err := ioutil.TempDir("/tmp", "glb-xdp")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer os.RemoveAll(tmpDir)
+
+	// prepare a socket where we can match systemd behaviour, waiting for
+	// the newly forked process to send READY.
+	readySock := tmpDir + "/ready.sock"
+	socketAddr := &net.UnixAddr{
+		Name: readySock,
+		Net:  "unixgram",
+	}
+	conn, err := net.ListenUnixgram("unixgram", socketAddr)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer conn.Close()
+
+	// we'll rexec with the same command and args that we were called with.
+	cmd := exec.Command(os.Args[0], os.Args[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Env = append(os.Environ(),
+		"NOTIFY_SOCKET=" + readySock,
+	)
+
+	err = cmd.Start()
+	if err != nil {
+		log.Fatalf("gracefulRestart: Failed to launch, error: %v", err)
+	}
+
+	fmt.Printf("New version of glb-director-xdp launched, waiting for READY signal...\n")
+
+	// wait for ready
+	var buf [1024]byte
+	n, err := conn.Read(buf[:])
```

Comment:
What happens if this exces and the new copy never comes up? Seems like you'd only get once chance at rolling over and things would be stuck here waiting for it and never get back into a state where it could receive another signal and try again. 

Should this have a timeout before giving up, printing a message, and going back into the normal routine waiting for another signal to try again?

Reply:
:+1: I've just added a timeout as well, which will give up and allow another attempt.